src/crypto/CryptStateOCB2.cpp

```diff
@@ -270,21 +270,39 @@ bool CryptStateOCB2::ocb_encrypt(const unsigned char *plain, unsigned char *encr
 	ZERO(checksum);
 
 	while (len > AES_BLOCK_SIZE) {
-		S2(delta);
-		XOR(tmp, delta, reinterpret_cast< const subblock * >(plain));
-		AESencrypt(tmp, tmp, &encrypt_key);
-		XOR(reinterpret_cast< subblock * >(encrypted), delta, tmp);
-		XOR(checksum, checksum, reinterpret_cast< const subblock * >(plain));
-
 		// Counter-cryptanalysis described in section 9 of https://eprint.iacr.org/2019/311
 		// For an attack, the second to last block (i.e. the last iteration of this loop)
 		// must be all 0 except for the last byte (which may be 0 - 128).
+		bool flipABit = false; // *plain is const, so we can't directly modify it
 		if (len - AES_BLOCK_SIZE <= AES_BLOCK_SIZE) {
 			unsigned char sum = 0;
 			for (int i = 0; i < AES_BLOCK_SIZE - 1; ++i) {
 				sum |= plain[i];
 			}
-			success &= sum != 0;
+			if (sum == 0) {
+				if (modifyPlainOnXEXStarAttack) {
+					// The assumption that critical packets do not turn up by pure chance turned out to be incorrect
+					// since digital silence appears to produce them in mass.
+					// So instead we now modify the packet in a way which should not affect the audio but will
+					// prevent the attack.
+					flipABit = true;
+				} else {
+					// This option still exists but only to allow us to test ocb_decrypt's detection.
+					success = false;
+				}
+			}
+		}
+
+		S2(delta);
+		XOR(tmp, delta, reinterpret_cast< const subblock * >(plain));
+		if (flipABit) {
+			*reinterpret_cast< unsigned char * >(tmp) ^= 1;
```

Comment:
Actually, my humble suggestion would be, defining a variable called
`char filpbit = 0` in the beginning and set it to 1 if flipping is needed.
This way, L298-L230 would directly turn to `*reinterpret_cast< unsigned char * >(tmp) ^= flipbit;` and the efficiency problem caused by the if statement would be saved. What do you think about this?

Reply:
I can't say much on the actual effects of such a change but if it does increase performance by removing if-branches that the optimizer doesn't get rid of anyways, it might be a good thing to do.

I think however there could be a comment as to which bit exactly is being flipped. Looking at this now I am wondering: Does the position of the flipped bit depend on the endianness of the machine this is running on? I think so right?
In that case we should directly use binary notation to get red of this dependence.
@Johni0702 was I mislead here?

Reply:
I'll put my faith into the compiler to apply that or a similar optimization itself. Not that two ifs will make much difference given what we had to go through to get here, at least on the client. And on the server I'd hope the CPU's branch predictor will get that jump right in the fast majority of cases (it'll always be false unless some client is using voice over TCP).

As for which bit is being flipped: Always the least significant bit of the first byte of the second to last block. We're dealing with a `char` pointer, so I don't see how it would be ambiguous (unless there's some part of the C++ standard which I'm unaware of).
I don't actually know if that `reinterpret_cast` is properly defined but an identical one already exists in the same method, so I've based my assumption on that one.

Reply:
You are absolutely right. We have no reason to put more effort into a patch of an obsolete thing. As long as it works, we'd better not bother it anymore.

Reply:
Well a `char` is also a number and therefore susceptible to being either big or little endian. And while I would assume that something like the Opus Codec does not depend on endianness (otherwise it wouldn't work across different machines that use different endiannesses) we have to make sure to flip the correct bit regardless of whether the system this has been compiled for uses big or little endian :thinking: 
At least I _think_ that this could be a problem...

Reply:
Endianness usually refers to the byte order in which multi-byte data is stored in memory (or some other medium). There is only one way to store a `char` since it is the smallest possible unit (except maybe for some really bizarre architectures).

Talking about which bit of that byte comes first makes no sense in most contexts and is pretty much an implementation detail of your CPU (and that's assuming its registers aren't three dimensional where "first" makes even less sense).
Those systems (e.g. compression algorithms, or serial ethernet) that need to have a specific bit-order (then called bit-endianness) usually specify upfront that the MSB is either first or last.
Most systems however, including IP/UDP and most architectures treat a byte (aka octet) as the smallest unit and all operations are defined relative to it, e.g. `^= 1` sets the LSB, regardless of where that LSB is actually located in the silicon / on the wire.

Having said that, for the fix it's completely irrelevant which bit(s) is set, as long as the result is no longer all 0. The Opus decoder on the other side does not read that byte anyway.

Reply:
Ah okay then I always misunderstood what endianness is about. Thanks for the explanation :+1: